### DISCUSSION TITLE
postgresql major version upgrade

### DISCUSSION THREAD

**Discussion by glehmann** [cloudnative-pg/cloudnative-pg#1176] [Ideas]
As noted in https://cloudnative-pg.io/documentation/1.17/rolling_update/

> Only upgrades for PostgreSQL minor releases are supported.

Is there a plan to add the update feature for a major release?

Would you be interested in a PR for this feature?
It seems to me it would be relatively simple to run `initDb`, `pg_upgrade --check` and `pg_upgrade --links` in an init container, in a job or from the controller before starting postgres. But well, that's a quite naive opinion, and it might be too simple to be true :-)

**Comment by sxd** [comment]
Hi @glehmann 

For now this can be handled by creating another PostgreSQL with the major version and replicate from there, probably it's not the best solution for now, but it's a solution, since it's required a few more process to upgrade a current instance and it's not that easy to handled the errors that could happen.

I'm moving this into a discussion where we can discuss the future of this possible feature.

**Reply by vladimirfx** [reply]
What kind of replication do you mean?

**Reply by sxd** [reply]
@vladimirfx when you use the cluster replication feature, you can read more here https://cloudnative-pg.io/documentation/1.18/replica_cluster/

**Reply by vladimirfx** [reply]
As I understand docs for replica cluster it uses physical replication type witch is requires same major Pg version. 
I can't understand how this is solves major upgrade task, sorry. Can you wrote short upgrade plan? Thank you

**Reply by jakubhajek** [reply]
I am also interested in that topic. Based on my testing the best option is to manually create logical replication in order to proceed with the major cluster version. Since it is logical replication, the target instance can be in a different PG version. 
However, that would be an excellent feature if we could orchestrate the major upgrade using the CNPG operator.

**Reply by vladimirfx** [reply]
We are trying manual pg_uprade using Postgres cluster hibernation feature.
Idea is simple:
- hibernate cluster
- attach cluster PV to special upgrade pod (image with several pg versions installed) 
- perform manual pg_upgrade
- adjust Pg image in hibernated cluster to target major version
- unhibernate cluster

**Reply by vladimirfx** [reply]
> We are trying manual pg_uprade using Postgres cluster hibernation feature. Idea is simple:
> 
> * hibernate cluster
> * attach cluster PV to special upgrade pod (image with several pg versions installed)
> * perform manual pg_upgrade
> * adjust Pg image in hibernated cluster to target major version
> * unhibernate cluster

Ofcose All of this can be automated.

**Reply by jakubhajek** [reply]
It sounds like a great solution, I will try it and share my experiences. Thanks for the inspiration, @vladimirfx :)

**Reply by gbartolini** [reply]
I agree it is an excellent feature. For the near future I won't be able to work directly on this (and I think most of my colleagues at EDB), but we can definitely supervise anyone who'd like to submit a PR.

I recall suggesting a similar procedure to the one @vladimirfx is trying, in the Slack chat, as a workaround. For pg_upgrade support, we need to create a special image in the postgres-containers repo (or a separate one) containing all versions of Postgres in a container. If anyone wants to try this, we can definitely help creating the repo.

**Reply by clrxbl** [reply]
I wouldn't mind giving the above a shot as I'd like to update from PostgreSQL 15 to 16 at some point. A separate repository seems like it would make the most sense considering e.g. postgis is separated already aswell.

That is assuming this is still necessary, cc: @gbartolini

**Comment by especially-relative** [comment]
Currently it seems like the state of the art is to make a new BG cluster
named ApplicationDatabase(2), which uses logical replication, then update
the application settings to point at the ApplicationDatabase(2) service? It
going to be an offline process either way but the current method just feels
very messy and manual unless I am missing something.

**Comment by gbartolini** [comment]
There’s actually the import feature which works fine. It doesn’t support zero cutover upgrades but it is definitely something that works for many use cases while we implement online migrations.
See https://cloudnative-pg.io/documentation/current/database_import/

**Reply by vladimirfx** [reply]
Yes, it is an absolutely working (and IMO recommended) approach for small databases (< 200 GB). For bigger databases, upgrade time becomes unacceptable because of dump/restore performance. So pg_upgrade method is very welcome.
Thank you for your response and elaborate software!

**Reply by scoutinho-randori** [reply]
@vladimirfx did it work for you?

**Reply by scoutinho-randori** [reply]
When i tried to hibernate and un-hibernate the cluster i got the following error ```
{"level":"info","ts":1680568494.3849115,"msg":"PGData already exists, can't overwrite","logging_pod":"cluster-fourteen-1"}```

**Reply by aelbarkani** [reply]
I agree with @vladimirfx. When the database is too big, pg_dump/pg_restore becomes too slow, so pg_upgrade would be more than welcome ! And many thanks for the operator, it's awesome !

**Reply by vladimirfx** [reply]
> @vladimirfx did it work for you?

Last year we had a successful upgrade in stage env but with many adjustments by hand. But recently I've tried upgrading to 17 and... operator (1.24.0) simply recreated cluster (!) after rehydrating (`cnpg.io/hibernation: off`) with this log:

```
{"level":"error","ts":"2024-10-04T07:49:50Z","msg":"Reconciler error","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"searchcitydb","namespace":"search-city"},"namespace":"search-city","name":"searchcitydb","reconcileID":"f548b48f-cc1c-48c0-a708-d2b9c6880254","error":"cannot reconcile required plugins: clusters.postgresql.cnpg.io \"searchcitydb\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:324\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222"}
{"level":"info","ts":"2024-10-04T07:49:50Z","msg":"Resource has been deleted","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"searchcitydb","namespace":"search-city"},"namespace":"search-city","name":"searchcitydb","reconcileID":"72d4a700-0de2-47aa-923e-d4c75372c5ca"}
{"level":"info","ts":"2024-10-04T07:49:50Z","logger":"cluster-resource","msg":"validate create","version":"v1","name":"searchcitydb","namespace":"search-city"}
{"level":"info","ts":"2024-10-04T07:49:50Z","msg":"Resource has been deleted","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"searchcitydb","namespace":"search-city"},"namespace":"search-city","name":"searchcitydb","reconcileID":"f75f951d-3966-4414-b15b-71541786792b"}
{"level":"info","ts":"2024-10-04T07:49:50Z","msg":"no orphan PVCs found, skipping the restored cluster reconciliation","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"searchcitydb","namespace":"search-city"},"namespace":"search-city","name":"searchcitydb","reconcileID":"33b28b6b-2942-4a4a-8494-ea5a5062998a"}
```

Maybe I do something wrong. I've to try later.

**Reply by vladimirfx** [reply]
> When i tried to hibernate and un-hibernate the cluster i got the following error ` {"level":"info","ts":1680568494.3849115,"msg":"PGData already exists, can't overwrite","logging_pod":"cluster-fourteen-1"}`

Looks like rehydration has not started at all because this error message indicates the bootstrap phase.

After upgrading I moved all CNPG-specific files to the new cluster AND after the new pod creation, I patched them by hand to a new container image (the operator used the old one instead of the new one set in cluster resource). As of now, it does not look like a production approach.

**Comment by clrxbl** [comment]
pg_upgrade support would be very welcome coming from the Crunchy Data Postgres Operator to cloudnative-pg with a PostgreSQL 15 cluster now that PostgreSQL 16 is out. It's really the only core feature that I'm missing in this operator right now.

The way Crunchy Data handles this:
- Create PGUpgrade custom resource that contains the upgrade image, major version numbers (from and to the version you're upgrading) and the cluster that should be targeted
- Shutdown/hibernate & annotate cluster to allow the above upgrade
- Operator now creates pods with the above upgrade image which just runs pg_upgrade
- PGUpgrade should now be completed & ready to start the cluster

https://access.crunchydata.com/documentation/postgres-operator/latest/guides/major-postgres-version-upgrade

**Reply by gbartolini** [reply]
The main issues I currently see with in place `pg_upgrade` (the way for example Crunchy has done) are:

- no way to rollback if something goes wrong (which might be fine if your database is 10GB as you can recover, but not ideal if your database is a few terabytes)
- no support for PostGIS

Our current plan is to revisit `pg_upgrade` support once we have implemented and completed our declarative enablement for very large databases that includes: volume snapshots backup & recovery, and tablespaces support.

In particular, as a tail for volume snapshot backup and recovery work we want to enable PVC cloning as a way to bootstrap a cluster. This will allow us to safely use PVC cloning (with copy on write) for in-place pg_upgrade usage as, in case of failure, we can simply rollback by using the previous volumes (untouched).

So ... we will eventually get there. Otherwise, a better approach (IMHO) is to rely on logical replication. Although we are not still there, please read the article I wrote a few weeks ago and that I have shared in this thread.

**Reply by clrxbl** [reply]
> no way to rollback if something goes wrong

I believe Crunchy Data stores their pgdata in the same PVC by major version number, e.g. /data/pg14 for PostgreSQL 14, /data/pg15 for PostgreSQL 15. So you could still rollback this way by using the same PVC.

I would prefer to do in-place upgrades as we can always restore from backups and I run my PostgreSQL cluster on baremetal without any form of block/network attached storage, so there isn't always space available for hosting another replica of the same database.

**Comment by wvengen** [comment]
Looking forward to any features in cloudnative-pg to help with this.

I'm currently using this very unpolished script to automate this, while preserving the resource name, without going the `pg_upgrade` route.
https://gist.github.com/wvengen/3cafc308c1737ca8bf3c08e0054e6e75

**Comment by gbartolini** [comment]
Did you look at this? https://www.enterprisedb.com/blog/current-state-major-postgresql-upgrades-cloudnativepg-kubernetes

**Reply by gbartolini** [reply]
This is an important update: https://www.gabrielebartolini.it/articles/2024/03/cloudnativepg-recipe-5-how-to-migrate-your-postgresql-database-in-kubernetes-with-~0-downtime-from-anywhere/

**Reply by rocket357** [reply]
I would love to see the ability to restore an "upgrade" cluster via volumeSnapshot from the source cluster (which would require the upgrade cluster to be the same version as the source), then upgrade the "upgrade" cluster in place, then continue with the logical replication approach (if you are taking a volume snapshot ~1/hour in the source cluster, it'll greatly reduce the load for logical replication for databases several hundred GB or potentially a few TB in size, even if the pg_upgrade fails or has to be done a few times).  This can all happen without touching the source cluster (aside from enabling logical replication and taking the volume snapshots, of course).  Then once the large "upgrade" cluster is finished sync'ing, you can sort out sequences and such, cut the application over, and go on about your merry way.  It'd be a *lot* faster than using logical replication for the entire db.

**Comment by isdinAghmir** [comment]
Hello guys, 

Is this subject forgotten ? 

It would be a great feature for us to have this pgupgrade service, which is already implemented in CrunchyData Operator: [PGUpgrade Service](https://access.crunchydata.com/documentation/postgres-operator/latest/guides/major-postgres-version-upgrade). 

This kind of upgrades is very SaaS/IaaS friendly since the developper creates an 'upgrade request'. 

Kind regards

**Reply by gbartolini** [reply]
Hi @isdinAghmir,

The topic is still on our radar. We’re tentatively targeting it for version 1.26, as outlined in the [roadmap](https://github.com/orgs/cloudnative-pg/projects/1), though this may shift depending on EDB’s priorities and commitments to some of our customers - EDB is my company and the main contributor to the project FYI.

I understand that other operators already have this feature.

However, CloudNativePG faces a unique challenge: everything must be handled declaratively. The "PGUpgrade" resource, by nature, is implicitly imperative. Another critical consideration is rollback capability, which is essential to align with the self-healing paradigm. We can't support scenarios where rollback requires a full recovery or lacks PostGIS support.

That said, I still believe that a logical replication approach—which offers nearly zero cutover time—is preferable. We’re already working on declarative publications and subscriptions, which are planned for 1.25. These will enable you to declaratively import or upgrade an existing database. I recommend reading the article I shared earlier, which covers these topics in more detail.

We’re making steady progress—just think about how far we’ve come in the last two years, and where we’ll be two years from now. I encourage you to stay optimistic. ;)

Thanks for your interest, by the way.

**Reply by bo0tzz** [reply]
Very exciting to see this, I'm already a huge fan of CNPG and this roadmap covers all the last pieces that were (in my opinion) missing. Thank you!

**Comment by FlosseJonsson** [comment]
We would really like to have the pg_upgrade feature in CloudNativePG or at least a detailed instruction how to do it with the hibernation feature. I have tried to find it with google but with no success. Can you provide a document with more details on this hibernation method?
/Anders

**Comment by yevon** [comment]
It would be awesome to have a native supported way to upgrade between major postgresql versions,

**Comment by sxd** [comment]
@yevon you haven't read the documentation and or the release note right ?

**Comment by sxd** [comment]
This was implemented in 1.26.0 closing
